### ISSUE TITLE
optionally disable NTP daemon installation+config

### ISSUE THREAD

**Issue by jsleeio** [kubernetes/kops#9307]
**1. Describe IN DETAIL the feature/behavior/change you would like to see.**

**TLDR**: we build and use our own AWS AMIs with all required packages already installed and would like a toggle that lets me skip some or all of the OS package checks/remediation. We prefer `chrony` over `ntp` for its security track record. The `ntp` package in Ubuntu 18.04 [has unpatched CVEs](https://people.canonical.com/~ubuntu-security/cve/pkg/ntp.html). Additionally we will upgrade to Ubuntu 20.04 and at that point we **definitely** don't want to [switch to `systemd-timesyncd`](https://github.com/kubernetes/kops/pull/9182)

**Previously**: operated clusters on custom AMIs built from Debian stretch and used `execContainer` hooks to replace `ntp` package with `chrony`. This worked fine, though it slowed `nodeup` a little. Looked like the below. I think it was derived from the `cachefilesd` hooks example in the cluster spec docs.

https://github.com/kubernetes/kops/blob/master/docs/cluster_spec.md#hooks 

```
  hooks:
  - before:
    - kubelet.service
    execContainer:
      command:
      - sh
      - -c
      - chroot /rootfs apt-get update && chroot /rootfs apt-get install -y chrony
        && chroot /rootfs cp /root/chrony.conf /etc/chrony/chrony.conf
      image: busybox
    name: replace-ntpd-with-chrony.service
    requires:
    - ntp.service
  - before:
    - kubelet.service
    manifest: |
      Type=oneshot
      ExecStart=/bin/systemctl restart chrony.service
    name: restart-chrony.service
    requires:
    - replace-ntpd-with-chrony.service
```

**Now**: operating clusters on custom AMIs built from Ubuntu 18.04. `apt-get` seems to require DBus to work at all, so the `execContainer` hack doesn't work anymore, and when I tried to install it with a `oneshot` Systemd unit, `apt-get` kept getting killed with a `SIGTERM` before it could complete. Probably I'm doing something wrong with the Systemd unit, but...

Really I'd prefer to not have to do any of this. My AMIs already have an NTP daemon installed and configured.

**2. Feel free to provide a design supporting your feature request.**

As mentioned above, I don't think the _default_ behaviour should change. The current defaults are good for users that don't feel like setting up their own AMI pipeline.

Maybe a list in the cluster (or instance group?) spec of `nodeup` tasks to skip? In our case we would skip _all_ the package-related tasks, as our AMI already has the required packages.

```
spec:
  nodeup:
    skipTasks:
      - ntp
      - packages
```

There is a precedent here: there's already a knob for skipping the Docker install.

https://github.com/kubernetes/kops/blob/master/docs/cluster_spec.md#skip-install

One problem with this approach is that users that do build their own cloud images and skip tasks like this would need to pay close attention to any subsequent changes in the prerequisites. I can see this increasing the support burden for `kops`, and that's not good.

**Comment by hakman** [comment]
Thanks for the feedback @jsleeio, very detailed. Will try sped some time thinking about this.
/assign

**Comment by hakman** [comment]
For packages, the solution would be to use `pdatePolicy: external`:
https://kops.sigs.k8s.io/getting_started/arguments/#updatepolicy

**Comment by jsleeio** [comment]
Apologies for the delay in replying here, and thanks for looking into my issue, @hakman. I think the results of my testing (below) agree with your comment.

I can confirm that setting `updatePolicy: external` does **not** have the effect I'm looking for, though it may still be an option we'll use, as we build new patched AMIs every day. My test, using Kops 1.17.0 and a Kubernetes 1.17.6 cluster:

1. built an Ubuntu 18.04 AMI with `chrony`
1. added `updatePolicy: external` to our cluster spec
1. removed the `hooks` section in cluster spec (that we were using to override the `ntp` config)
1. updated instance group specs with new AMI ID
1. `kops update cluster --yes`
1. rolled the `bastions` and `nodes` instance groups

After these changes, had a look at the new instances:

* new bastion has `chrony` configured per the new AMI
* new nodes have `ntp` configured per `kops` nodeup

On a `nodes` instance, `/var/log/apt/history.log` shows only this `apt-get` activity after boot:

```
Start-Date: 2020-06-14  12:41:50
Commandline: apt-get install --yes --no-install-recommends ntp
Install: ntp:amd64 (1:4.2.8p10+dfsg-5ubuntu7.1), libopts25:amd64 (1:5.18.12-4, automatic)
Remove: chrony:amd64 (3.2-4ubuntu4.4)
End-Date: 2020-06-14  12:41:55
```

**Comment by hakman** [comment]
Probably there is a confusion. Please check the description of `updatePolicy`. This will help with not installing packages that you already have installed in you custom AMI, not all of them. Providing an option to completely disable doesn't seem like a good idea.

There is no option to disable NTP yet, but there was a plan to add various options to control NTP, like enable/disable, host(s) and maybe even daemon. This is pretty easy to implement and I would be quite happy to review such a changes, in case someone has more time than me to get this done.
